hub/models.py

```diff
@@ -3303,6 +3330,480 @@ def get_import_data(self):
+            client_config=settings.GOOGLE_SHEETS_CLIENT_CONFIG, scopes=cls.SCOPES
+        )
+
+    @classmethod
+    def from_oauth_redirect_success(cls, redirect_success_url: str, **kwargs):
+        """
+        Create an instance of the class, converting the redirect_success_url
+        into oauth credentials with the Google API.
+        """
+        flow = cls.oauth_flow()
+        flow.redirect_uri = redirect_success_url.split("?")[0]
+        token = flow.fetch_token(authorization_response=redirect_success_url)
+        oauth_credentials = json.dumps(
+            {
+                "access_token": token["access_token"],
+                "refresh_token": token["refresh_token"],
+                "client_id": flow.client_config["client_id"],
+                "client_secret": flow.client_config["client_secret"],
+                "scopes": token["scope"],
+            }
+        )
+        return cls(oauth_credentials=oauth_credentials, **kwargs)
+
+    @classmethod
+    def get_deduplication_field_names(cls) -> list[str]:
+        return ["spreadsheet_id", "sheet_name"]
+
+    @classmethod
+    def authorization_url(cls, redirect_url: str) -> str:
+        flow = cls.oauth_flow()
+        flow.redirect_uri = redirect_url
+        authorization_url, state = flow.authorization_url(
+            access_type="offline", include_granted_scopes="true", prompt="consent"
+        )
+        return authorization_url
+
+    @cached_property
+    def api(self):
+        credentials = Credentials.from_authorized_user_info(
+            json.loads(self.oauth_credentials)
+        )
+        return googleapiclient.discovery.build("sheets", "v4", credentials=credentials)
+
+    @cached_property
+    def spreadsheets(self) -> googleapiclient.discovery.Resource:
+        return self.api.spreadsheets()
+
+    @cached_property
+    def spreadsheet(self):
+        return self.spreadsheets.get(spreadsheetId=self.spreadsheet_id).execute()
+
+    @property
+    def sheet(self):
+        return self.get_sheet(sheet_name=self.sheet_name)
+
+    @property
+    def lookup_sheet(self):
+        return self.get_sheet(sheet_name="MAPPED_LOOKUP_SHEET")
+
+    @property
+    def webhook_sheet(self):
+        return self.get_sheet(sheet_name="MAPPED_WEBHOOK_SHEET")
+
+    @cached_property
+    def headers(self) -> list[str]:
+        result = (
+            self.spreadsheets.values()
+            .get(spreadsheetId=self.spreadsheet_id, range=f"{self.sheet_name}!1:1")
+            .execute()
+        )
+        return result["values"][0]
+
+    def get_sheet(self, sheet_name: str):
+        return next(
+            (
+                sheet
+                for sheet in self.spreadsheet["sheets"]
+                if sheet["properties"]["title"] == sheet_name
+            ),
+            None,
+        )
+
+    def remote_url(self) -> str:
+        sheet = self.sheet
+        return f"{self.spreadsheet['spreadsheetUrl']}?gid=${sheet['properties']['sheetId']}"
```

Comment:
Wonder if we could construct this a bit more resiliently using `urllib3` to build the params etc.